resources/providers/awslib/aws.go

```diff
@@ -18,20 +18,85 @@
 package awslib
 
 import (
-	"github.com/aws/aws-sdk-go-v2/aws"
+	"context"
+	"fmt"
+	awssdk "github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"sync"
 )
 
-type Config struct {
-	Config aws.Config
+const DefaultRegion = "us-east-1"
+
+var (
+	instance *singleton
+	once     = &sync.Once{}
+)
+
+type singleton struct {
+	regions []string
 }
 
-type Metadata struct {
-	Resource string
-	Arn      string
+type Config struct {
+	Config awssdk.Config
 }
 
 type AwsResource interface {
 	GetResourceArn() string
 	GetResourceName() string
 	GetResourceType() string
 }
+
+type AWSCommonUtil interface {
+	DescribeRegions(ctx context.Context, params *ec2.DescribeRegionsInput, optFns ...func(*ec2.Options)) (*ec2.DescribeRegionsOutput, error)
+}
+
+type MultiRegionWrapper[T any] struct {
+	Clients map[string]T
+}
+
+// CreateMultiRegionClients is a utility function that is used to create a map of client instances of a given type T for multiple regions.
+func CreateMultiRegionClients[T any](client AWSCommonUtil, cfg awssdk.Config, factory func(cfg awssdk.Config) T, log *logp.Logger) *MultiRegionWrapper[T] {
+	var clientsMap = make(map[string]T, 0)
+	for _, region := range getRegions(client, log) {
+		cfg.Region = region
+		client := factory(cfg)
+		clientsMap[region] = client
```

Comment:
As we talked about on Thursday I think this map should hold our providers instead of the raw AWS clients. It will give us more flexibility in the future. (There's nothing wrong with the function implementation just the usage)